File: leetcode/980_Unique_Paths_III.py
from typing import List
from unittest import TestCase

class Solution:
    def uniquePathsIII(self, grid: List[List[int]]) -> int:
        n, m = len(grid), len(grid[0])
        start, empty_n= (0, 0), 0
        in_stack = [[False] * m for _ in  range(n)]
        path_cnt, in_stack_cnt = 0, 0
        for x in range(n):
            for y in range(m):
                if grid[x][y] == 0:
                    empty_n += 1
                elif grid[x][y] == 1:
                    start = (x, y)
        
        def dfs(x: int, y: int):
            nonlocal in_stack_cnt
            nonlocal in_stack
            nonlocal path_cnt
            in_stack[x][y] = True
            in_stack_cnt += 1
            for dx, dy in ((0, 1), (1, 0), (0, -1), (-1, 0)):
                x1, y1 = x + dx, y + dy
                if 0 <= x1 < n and 0 <= y1 < m:
                    if grid[x1][y1] == 0 and not in_stack[x1][y1]:
                        dfs(x1, y1)
                    elif grid[x1][y1] == 2 and in_stack_cnt == empty_n + 1:
                        path_cnt += 1
            in_stack[x][y] = False
            in_stack_cnt -= 1

        dfs(*start)
        return path_cnt

class UniquePathsIIITest(TestCase):
    def test_1(self):
        self.assertEqual(2, Solution().uniquePathsIII([[1,0,0,0],[0,0,0,0],[0,0,2,-1]]))

    def test_2(self):
        self.assertEqual(4, Solution().uniquePathsIII([[1,0,0,0],[0,0,0,0],[0,0,0,2]]))

    def test_3(self):
        self.assertEqual(0, Solution().uniquePathsIII([[0,1],[2,0]]))

File: leetcode/980_Unique_Paths_III.rs
struct Solution {}

impl Solution {
    pub fn unique_paths_iii(mut grid: Vec<Vec<i32>>) -> i32 {
        fn dfs(x: usize, y: usize, grid: &Vec<Vec<i32>>, n: usize, m: usize, empty_cnt: u32, in_stack: &mut Vec<Vec<bool>>, in_stack_cnt: &mut u32, path_cnt: &mut i32)  {
            in_stack[x][y] = true;
            *in_stack_cnt += 1;
            for (x1, y1) in [(0, 1), (1, 0), (0, -1), (-1, 0)].map(|(dx, dy)| (x as isize + dx, y as isize + dy)) {
                if  0 <= x1 && x1 < n as isize && 0 <= y1 && y1 < m as isize  {
                    let (x1, y1) = (x1 as usize, y1 as usize);
                    if grid[x1][y1] == 0 && ! in_stack[x1][y1] {
                        dfs(x1, y1, grid, n, m, empty_cnt, in_stack, in_stack_cnt, path_cnt);               
                    } else if grid[x1][y1] == 2 && *in_stack_cnt == empty_cnt + 1 {
                        *path_cnt += 1;
                    }

                }
            }            
            in_stack[x][y] = false;
            *in_stack_cnt -= 1;
        }

        let (n, m) = (grid.len(), grid[0].len());
        let mut path_cnt = 0;
        let mut empty_cnt = 0;
        let (mut start_x, mut start_y) = (0, 0);
        for x in 0..n {
            for y in 0..m {
                if grid[x][y] == 0 {
                    empty_cnt += 1;
                } else if grid[x][y] == 1 {
                    start_x = x;
                    start_y = y;
                }
            }
        }
        let mut in_stack = vec![vec![false; m]; n];
        let mut in_stack_cnt = 0;
        dfs(start_x, start_y, &mut grid, n, m, empty_cnt, &mut in_stack, &mut in_stack_cnt, &mut path_cnt);
        path_cnt
    }
}

#[test]
fn test_1() {
    assert_eq!(2, Solution::unique_paths_iii(vec![vec![1,0,0,0],vec![0,0,0,0],vec![0,0,2,-1]]));
}

#[test]
fn test_2() {
    assert_eq!(4, Solution::unique_paths_iii(vec![vec![1,0,0,0],vec![0,0,0,0],vec![0,0,0,2]]));
}

#[test]
fn test_3() {
    assert_eq!(0, Solution::unique_paths_iii(vec![vec![0,1],vec![2,0]]));
}